File: ckgame-java/src/main/java/com/ckgame/world/Title.java
package com.ckgame.world;

import com.ckgame.core.Constants;
import com.ckgame.core.TitleTier;
import com.ckgame.politics.RealmLaw;

import java.util.ArrayList;
import java.util.List;

/** 头衔（领地）。 */
public final class Title {
    public int id;
    public String name;
    public TitleTier tier;
    public String adjective = "";
    public int holder = Constants.NONE_ID;
    public int deJureLiege = Constants.NONE_ID;
    public int deFactoLiege = Constants.NONE_ID;
    public List<Integer> deJureVassals = new ArrayList<>();
    public List<Integer> deFactoVassals = new ArrayList<>();
    public int capital = Constants.NONE_ID;
    public List<Integer> counties = new ArrayList<>();
    public double creationCost = 0.0;
    public boolean destroyable = false;
    public RealmLaw realmLaw = RealmLaw.feudalDefault();

    public Title(int id, String name, TitleTier tier) {
        this.id = id;
        this.name = name;
        this.tier = tier;
    }

    public static Title newTitle(int titleId, String name, TitleTier tier) {
        Title t = new Title(titleId, name, tier);
        t.adjective = name + "的";
        t.creationCost = tier.creationCost();
        t.destroyable = tier.compareTo(TitleTier.DUCHY) >= 0;
        return t;
    }

    public boolean isHeld() {
        return holder != Constants.NONE_ID;
    }

    public void setHolder(int holder) {
        this.holder = holder;
    }

    public void clearHolder() {
        this.holder = Constants.NONE_ID;
    }
}

File: ckgame-ts/src/world/Title.ts
import { NONE_ID } from '../core/Constants.js';
import { TitleTier, tierCreationCost, tierIsDestroyable } from '../core/TitleTier.js';
import { RealmLaw } from './Laws.js';

/** 头衔（领地）。 */
export class Title {
  id: number;
  name: string;
  tier: TitleTier;
  adjective: string = '';
  holder: number = NONE_ID;
  deJureLiege: number = NONE_ID;
  deFactoLiege: number = NONE_ID;
  deJureVassals: number[] = [];
  deFactoVassals: number[] = [];
  capital: number = NONE_ID;
  counties: number[] = [];
  creationCost: number = 0;
  destroyable: boolean = false;
  realmLaw: RealmLaw = RealmLaw.feudalDefault();

  constructor(id: number, name: string, tier: TitleTier) {
    this.id = id;
    this.name = name;
    this.tier = tier;
  }

  static newTitle(titleId: number, name: string, tier: TitleTier): Title {
    const t = new Title(titleId, name, tier);
    t.adjective = name + '的';
    t.creationCost = tierCreationCost(tier);
    t.destroyable = tierIsDestroyable(tier);
    return t;
  }

  isHeld(): boolean {
    return this.holder !== NONE_ID;
  }

  setHolder(holder: number): void {
    this.holder = holder;
  }

  clearHolder(): void {
    this.holder = NONE_ID;
  }
}
